### PR TITLE
Add option to preserve newlines in XML documents

### PR DESCRIPTION
**Pull request by kwokcb**
Fixes:  #1316

* Add a option to `pugixml` to create node_newline` DOM nodes while parsing white space. Stores as CR/LF in it's value member.
* Add option to `XmlReadOptions.readNewlines` for XML read options.
* Add in a new `NewlineElement` to store new elements corresponding to `blank_node` DOM nodes. Stores
a CR/LF character in it's doc string. 
* Add Python wrapper `PyNewlineElement`.

Testing:
* Add new unit test to read chess set example file (which has comments + empty newlines) and write it back out to compare against the original. 
* Modify `mxupdate.py `so that it now enables blank line preservation (run in unit tests). Note that chess file
was run through `mxupdate` once to get new "baseline" file for unit tests.

### CODE REVIEW

**Comment by kwokcb** [comment]
@jstone-lucasfilm . This will now allow preservation of blank lines but as noted is a modification to external code (`pugixml`).

File: source/MaterialXFormat/External/PugiXML/pugixml.cpp [hunk 1]
```diff
@@ -3419,7 +3419,31 @@ PUGI__NS_BEGIN
...
+					// Add a new node per new blank line, while eating whitespace
+					if (PUGI__OPTSET(parse_blank_lines))
+					{
+						if (PUGI__IS_CHARTYPE(*s, ct_space))
+						{
+							unsigned int lineCount = 0;
+							while (PUGI__IS_CHARTYPE(*s, ct_space))
+							{
+								if (s[0] == '\n')
+								{
```
Comment (kwokcb) [inline source/MaterialXFormat/External/PugiXML/pugixml.cpp:3431]: The logic is to keep a EOL count while parsing whitespace. Then adding blank line nodes  as needed.

File: source/MaterialXFormat/XmlIo.cpp [hunk 2]
```diff
@@ -63,11 +69,17 @@ void elementFromXml(const xml_node& xmlNode, ElementPtr elem, const XmlReadOptio
...
 
         // Handle the interpretation of XML comments.
-        if (readOptions && readOptions->readComments && category.empty())
+        if (readOptions && readOptions->readComments && category.empty() && !addBlank)
         {
             child = elem->changeChildCategory(child, CommentElement::CATEGORY);
             child->setDocString(xmlChild.value());
         }
+
+        if (readOptions && readOptions->readBlankLines && category.empty() && addBlank)
```
Comment (kwokcb) [inline source/MaterialXFormat/XmlIo.cpp:78]: Create blank line elements on read.

File: source/MaterialXFormat/XmlIo.cpp [hunk 3]
```diff
@@ -123,6 +135,14 @@ void elementToXml(ConstElementPtr elem, xml_node& xmlNode, const XmlWriteOptions
             }
         }
 
+        // Write blank lines.
+        if (child->getCategory() == BlankLineElement::CATEGORY)
```
Comment (kwokcb) [inline source/MaterialXFormat/XmlIo.cpp:139]: Add blank lines on write.

File: source/MaterialXView/Viewer.cpp [hunk 4]
```diff
@@ -1214,6 +1214,8 @@ void Viewer::loadDocument(const mx::FilePath& filename, mx::DocumentPtr librarie
 {
     // Set up read options.
     mx::XmlReadOptions readOptions;
+    readOptions.readComments = true;
```
Comment (kwokcb) [inline source/MaterialXView/Viewer.cpp:1217]: I think it's useful to always turn these on to try and always preserve the original content.
Maybe also do this for the Graph Editor?

File: source/MaterialXFormat/External/PugiXML/pugixml.cpp [hunk 5]
```diff
@@ -3419,7 +3419,31 @@ PUGI__NS_BEGIN
 				{
 					mark = s; // Save this offset while searching for a terminator.
 
-					PUGI__SKIPWS(); // Eat whitespace if no genuine PCDATA here.
+					// Add a new node per new blank line, while eating whitespace
```
Comment (jstone-lucasfilm) [inline source/MaterialXFormat/External/PugiXML/pugixml.cpp:3422]: Since we're making changes to the core PugiXML code, and will need to maintain them across future revisions, we should bracket these changes with MaterialX Begin/End comments, in the spirit of the single-line MaterialX comments found here:

https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/MaterialXFormat/External/PugiXML/pugixml.cpp#L3925

File: source/MaterialXView/Viewer.cpp (hunk 4 above)
Comment (jstone-lucasfilm) [inline source/MaterialXView/Viewer.cpp:1217]: I'd recommend omitting these edits from the viewer, where we don't have the expectation that comments and blank lines in the original document will be maintained.

**Review by jstone-lucasfilm — COMMENTED** [review]
This looks really promising, @kwokcb, and I've written up some minor recommendations for revisions.

**Comment by jstone-lucasfilm** [comment]
This looks very close to the mark, @kwokcb, and the one issue is like to resolve before merging is the indentation of newlines.  Because this new logic adds indentation to generated newlines, a round trip of a document through mxupdate.py doesn't quite match the original.

**Review by jstone-lucasfilm — COMMENTED** [review]
This new feature is coming along well, @kwokcb, but I can see two issues that still need to be resolved before this pull request can move forward:

1. We're not yet handling the new PugiXML node in functions such as `node_output_simple`, so PugiXML triggers internal asserts when a document containing newlines is written in a Debug build.  These asserts are invisible in Release builds, because PugiXML silently ignores them in this context.
2. PugiXML writes out stored newlines with _indentation_, so we don't get matching results when a document containing newlines is roundtripped through a tool such as `mxupdate.py`.

**Comment by kwokcb** [comment]
Hi @jstone-lucasfilm,
I'll take a look at the issues you mentioned. Thanks for being so thorough on this.

**Comment by kwokcb** [comment]
@jstone-lucasfilm. I ran this through debug and seems that the missing support is where you mentioned. Have fixed so newline nodes won't get indented so that `mxupdate.py` (or any other write back) won't introduce indented newlines.

**Comment by kwokcb** [comment]
This is a bit strange as I don't see any platform specific code and the previous test file worked.
I tried resaving the chess set on Mac and the entire file has whitespace differences. Retesting passes with the new version. Maybe it's an issue with the readFile() utility ?

**Comment by kwokcb** [comment]
Guessing it was a file specific formatting issue -- not sure how that file was created but seems it was never saved through PugiXML write due to all the formatting differences. 

I ran the upgrade tests and the formatting also seems fine for those files. e.g. the 1.36 to 1.37 resulting file has correctly preserved indentation.
```xml
<?xml version="1.0"?>
<materialx version="1.38">

  <!-- invert matrix to invertmatrix -->
  <invertmatrix name="invert1" type="matrix33">
    <input name="in" type="matrix33" value="1.0, 0.0, 0.0,  0.0, 1.0, 0.0,  0.0, 0.0, 1.0" />
  </invertmatrix>
  <invertmatrix name="invert2" type="matrix44">
    <input name="in" type="matrix44" value="1.0, 0.0, 0.0, 0.0,  0.0, 1.0, 0.0, 0.0,  0.0, 0.0, 1.0, 0.0,  0.0, 0.0, 0.0, 1.0" />
  </invertmatrix>

  <!-- no change for non-matrix invert variants -->
  <invert name="invert_no_change_float" type="float">
    <input name="in" type="float" value="0.5" />
```

**Comment by kwokcb** [comment]
@jstone-lucasfilm . I've done what I think is a "final" sanity check by running `mxupdate` on all `libraries` and `resources` files and run through regression testing..  I checked updates for stdlib only for the PR -- there are no changes to pbrlib.

This mostly found existing formatting issue fixes, but for `bxdf` there is the case such as for the `lama` and `standardsurface` where the `doc` attribute has been indented on another line so will change. This is purely due to the existing element formatting logic.

File: libraries/stdlib/stdlib_defs.mtlx [hunk 6]
```diff
@@ -1004,7 +1004,7 @@
     <input name="type" type="integer" uiname="Noise Type" uifolder="Common" uisoftmin="0" uisoftmax="3" value="0" enum="Perlin,Cell,Worley,Fractal" enumvalues="0,1,2,3" doc="Menu to select the type of noise: Perlin, Cell, Worley, or Fractal. Default is Perlin." />
     <output name="out" type="float" />
   </nodedef>
-  
+
```
Comment (kwokcb) [inline libraries/stdlib/stdlib_defs.mtlx:1007]: Indented empty lines are not preserved by design as spaces and tabs are not preserved.

File: libraries/stdlib/genmdl/stdlib_genmdl_impl.mtlx [hunk 7]
```diff
@@ -269,8 +269,10 @@
   <implementation name="IM_divide_vector3FA_genmdl" nodedef="ND_divide_vector3FA" sourcecode="{{in1}} / {{in2}}" target="genmdl" />
   <implementation name="IM_divide_vector4_genmdl" nodedef="ND_divide_vector4" sourcecode="{{in1}} / {{in2}}" target="genmdl" />
   <implementation name="IM_divide_vector4FA_genmdl" nodedef="ND_divide_vector4FA" sourcecode="{{in1}} / {{in2}}" target="genmdl" />
-  <implementation name="IM_divide_matrix33_genmdl" nodedef="ND_divide_matrix33" sourcecode="{{in1}}" target="genmdl" /> <!-- TODO: Implement properly -->
-  <implementation name="IM_divide_matrix44_genmdl" nodedef="ND_divide_matrix44" sourcecode="{{in1}}" target="genmdl" /> <!-- TODO: Implement properly -->
+  <implementation name="IM_divide_matrix33_genmdl" nodedef="ND_divide_matrix33" sourcecode="{{in1}}" target="genmdl" />
```
Comment (kwokcb) [inline libraries/stdlib/genmdl/stdlib_genmdl_impl.mtlx:272]: The existing writing logic does no preserve comments on the same line.

**Review by jstone-lucasfilm — COMMENTED** [review]
This looks great to me, @kwokcb, though I would recommend that we apply our formatting updates to just a single MaterialX file for now, so that we're not combining both new features and wide-ranging formatting updates in a single pull request.  Perhaps we can apply the formatting changes to just the Chess Set for now, and leave the updates to our data libraries for a future pull request?

**Comment by kwokcb** [comment]
Sounds good. I've reverted the non chess files.

**Review by jstone-lucasfilm — APPROVED** [review]
This looks good to me, thanks @kwokcb!
